cache/cache_entry_stats.h

```diff
@@ -0,0 +1,152 @@
+// column families or multiple DBs sharing a Cache. We employ a few
+// mitigations:
+// * Only one collector for a particular kind of Stats is alive
+// for each Cache. This is guaranteed using the Cache itself to hold
+// the collector.
+// * A mutex ensures only one thread is gathering stats for this
+// collector.
+// * The most recent gathered stats are saved and simply copied to
+// satisfy requests within a time window (default: 3 minutes) of
+// completion of the most recent stat gathering.
+//
+// Template parameter Stats must be copyable and trivially constructable,
+// as well as...
+// concept Stats {
+//   // Notification before applying callback to all entries
+//   void BeginCollection(Cache*, SystemClock*, uint64_t start_time_micros);
+//   // Get the callback to apply to all entries. `callback`
+//   // type must be compatible with Cache::ApplyToAllEntries
+//   callback GetEntryCallback();
+//   // Notification after applying callback to all entries
+//   void EndCollection(Cache*, SystemClock*, uint64_t end_time_micros);
+//   // Notification that a collection was skipped because of
+//   // sufficiently recent saved results.
+//   void SkippedCollection();
+// }
+template <class Stats>
+class CacheEntryStatsCollector {
+ public:
+  // Gathers stats and saves results into `stats`
+  void GetStats(Stats *stats, int maximum_age_in_seconds = 180) {
+    // Waits for any pending reader or writer (collector)
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // Maximum allowed age is nominally given by the parameter
+    uint64_t max_age_micros =
+        static_cast<uint64_t>(std::min(maximum_age_in_seconds, 0)) * 1000000U;
+    // But we will re-scan more frequently if it means scanning < 1%
+    // of the time and no more than once per second.
+    max_age_micros = std::min(
+        max_age_micros,
+        std::max(uint64_t{1000000},
+                 100U * (last_end_time_micros_ - last_start_time_micros_)));
+
+    uint64_t start_time_micros = clock_->NowMicros();
+    if ((start_time_micros - last_end_time_micros_) > max_age_micros) {
+      last_start_time_micros_ = start_time_micros;
+      saved_stats_.BeginCollection(cache_, clock_, start_time_micros);
+
+      cache_->ApplyToAllEntries(saved_stats_.GetEntryCallback(), {});
+
+      uint64_t end_time_micros = clock_->NowMicros();
+      last_end_time_micros_ = end_time_micros;
+      saved_stats_.EndCollection(cache_, clock_, end_time_micros);
+    } else {
+      saved_stats_.SkippedCollection();
+    }
+    // Copy to caller
+    *stats = saved_stats_;
+  }
+
+  Cache *GetCache() const { return cache_; }
+
+  // Gets or creates a shared instance of CacheEntryStatsCollector in the
+  // cache itself, and saves into `ptr`. This shared_ptr will hold the
+  // entry in cache until all refs are destroyed.
+  static Status GetShared(Cache *cache, SystemClock *clock,
+                          std::shared_ptr<CacheEntryStatsCollector> *ptr) {
```

Comment:
shared_ptr is copyable. It's not needed here and now, but could be in the future. And I consider the fact that the CacheEntryStatsCollector is managed in the Cache to be an implementation detail.

Though thinking about it now, I am expecting all references to CacheEntryStatsCollector to be cleaned up before the Cache itself, only because I don't really have access to a shared_ptr version of Cache to reference from CacheEntryStatsCollector. This is a bit of implementation detail leakage.

Reply:
Yeah, what kind of threw me off was that the ability to copy the `shared_ptr` is not used, so there is actually no shared ownership of the cache handle. Hiding the fact that the collector lives in the cache is a good idea; there's a bit of overhead related to the `shared_ptr` but I think this fine since this is not a hot path.